Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp

```diff
@@ -0,0 +1,262 @@
+        }
+
+        m_solution[ilev]->setVal(0.);
+
+        mlpoisson.setLevelBC(ilev, m_solution[ilev].get());
+
+        MLMG mlmg(mlpoisson);
+        mlmg.setMaxIter(m_max_iter);
+        mlmg.setMaxFmgIter(m_max_fmg_iter);
+        mlmg.setVerbose(m_verbose);
+        mlmg.setBottomVerbose(m_bottom_verbose);
+        mlmg.solve({m_solution[ilev].get()}, {m_source[ilev].get()}, m_rtol, m_atol);
+
+        // Synchronize the ghost cells, do halo exchange
+        ablastr::utils::communication::FillBoundary(*(m_solution[ilev].get()),
+                                                m_solution[ilev]->nGrowVect(),
+                                                WarpX::do_single_precision_comms,
+                                                geom[ilev].periodicity());
+    }
+}
+
+void
+ProjectionDivCleaner::setSourceFromBfield ()
+{
+    // Get WarpX object
+    auto & warpx = WarpX::GetInstance();
+    const auto& geom = warpx.Geom();
+
+    // This function will compute divB and add to the source multifab
+    for (int ilev = 0; ilev < m_levels; ++ilev)
+    {
+        // Zero out source multifab
+        m_source[ilev]->setVal(0.0);
+
+        // Grab B-field multifabs at this level
+        std::array<const amrex::MultiFab* const, 3> Bfield =
+            warpx.getFieldPointerArray(warpx::fields::FieldType::Bfield_aux, ilev);
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+        for (MFIter mfi(*(m_source[ilev].get()), TilingIfNotGPU()); mfi.isValid(); ++mfi)
+        {
+            // Grab references to B field arrays for this grid/tile
+            amrex::Array4<Real const> const& Bx_arr = Bfield[0]->const_array(mfi);
+            amrex::Array4<Real const> const& By_arr = Bfield[1]->const_array(mfi);
+            amrex::Array4<Real const> const& Bz_arr = Bfield[2]->const_array(mfi);
+
+            // Extract stencil coefficients
+            Real const * const AMREX_RESTRICT coefs_x = m_stencil_coefs_x.dataPtr();
+            auto const n_coefs_x = static_cast<int>(m_stencil_coefs_x.size());
+            Real const * const AMREX_RESTRICT coefs_y = m_stencil_coefs_y.dataPtr();
+            auto const n_coefs_y = static_cast<int>(m_stencil_coefs_y.size());
+            Real const * const AMREX_RESTRICT coefs_z = m_stencil_coefs_z.dataPtr();
+            auto const n_coefs_z = static_cast<int>(m_stencil_coefs_z.size());
+
+            amrex::Array4<Real> const& src_arr = m_source[ilev].get()->array(mfi);
+
+            amrex::ParallelFor(mfi.tilebox(),
+            [=] AMREX_GPU_DEVICE (int i, int j, int k)
+            {
+                src_arr(i,j,k) -= CartesianYeeAlgorithm::DownwardDx(Bx_arr, coefs_x, n_coefs_x, i+1, j, k);
+                src_arr(i,j,k) -= CartesianYeeAlgorithm::DownwardDy(By_arr, coefs_y, n_coefs_y, i, j+1, k);
+                src_arr(i,j,k) -= CartesianYeeAlgorithm::DownwardDz(Bz_arr, coefs_z, n_coefs_z, i, j, k+1);
```

Comment:
Wouldn't the index stepping here cause issues in 1d/2d? Is there a reason not to use `UpwardD{x/y/z}` instead of `DownwardD{x/y/z}`?

Reply:
I had some issue with the prototype of UpwardD. I will see if I can adjust it so I don't need to explicitly index.

Reply:
This should now be moot as it has been removed in favor of using existing ComputeDivB.